File: 2015/cpp/day3/twosantas.cpp
#include <iostream>
#include <fstream>
#include <stdexcept>  // For throwing exceptions
#include <utility>    // For std::pair
#include <set>        // For std::set

std::pair<int, int> moveDirection(char direction, int x, int y) {
    switch (direction) {
        case '<': // Move left
            x -= 1;
            break;
        case '>': // Move right
            x += 1;
            break;
        case '^': // Move up
            y += 1;
            break;
        case 'v': // Move down
            y -= 1;
            break;
        default:
            throw std::invalid_argument("Invalid direction: " + std::string(1, direction));  // Throw exception for invalid direction
    }
    // Return the updated coordinates as a pair
    return std::make_pair(x, y);
}

int main() {
    std::ifstream file("input.txt");

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file." << std::endl;
        return 1;
    }
    int moves = 0;
    int santa_x = 0, santa_y = 0;
    int robosanta_x = 0, robosanta_y = 0;
    bool is_santa = moves % 2 == 1;
    // Create a set to store unique coordinates (visited positions)
    std::set<std::pair<int, int>> visited;

    // Add the initial starting position to the set
    visited.insert(std::make_pair(santa_x, santa_y));

    char ch;
    while (file.get(ch)) {
        moves += 1;
        is_santa = moves % 2 == 1;

        if (is_santa) {
            std::pair<int, int> newCoordinates = moveDirection(ch, santa_x, santa_y);
            santa_x = newCoordinates.first;
            santa_y = newCoordinates.second;
            visited.insert(newCoordinates);
        } else {
            std::pair<int, int> newCoordinates = moveDirection(ch, robosanta_x, robosanta_y);
            robosanta_x = newCoordinates.first;
            robosanta_y = newCoordinates.second;
            visited.insert(newCoordinates);
        }

    }
    std::cout << "after " << moves << " moves, " << visited.size() << " houses were visited" << std::endl;
    file.close();
    return 0;
}

File: 2015/java/aoc-2015/src/main/java/day3/TwoSantas.java
package day3;

import java.io.FileReader;
import java.io.IOException;
import java.lang.String;
import java.util.HashSet;
import java.util.Set;

public class TwoSantas {
    public static int[] moveDirection(char direction, int x, int y) {
        switch (direction) {
            case '<': // Move left
                x -= 1;
                break;
            case '>': // Move right
                x += 1;
                break;
            case '^': // Move up
                y += 1;
                break;
            case 'v': // Move down
                y -= 1;
                break;
            default:
                throw new IllegalArgumentException("Invalid direction: " + direction);
        }
        // Return the updated coordinates
        return new int[]{x, y};
    }

    public static String encode(int[] coordinates) {
        return coordinates[0] + "_" + coordinates[1];
    }

    public static void main(String[] args) throws IOException {
        FileReader inputStream = null;
        Set<String> visited = new HashSet<>();
        int moves = 0;
        int[] santa = {0, 0};
        int[] robosanta = {0, 0};
        boolean is_santa;
        visited.add(encode(santa));

        try {
            inputStream = new FileReader("src/main/resources/day3/input.txt");
            int c;
            while ((c = inputStream.read()) != -1) {
                moves += 1;
                is_santa = moves % 2 == 1;
                if (is_santa) {
                    santa = moveDirection((char) c, santa[0], santa[1]);
                    visited.add(encode(santa));
                } else {
                    robosanta = moveDirection((char) c, robosanta[0], robosanta[1]);
                    visited.add(encode(robosanta));
                }
            }
            System.out.printf("houses visited after %d moves is %d%n", moves, visited.size());
        } finally {
            if (inputStream != null) {
                inputStream.close();
            }
        }
    }    
}
